### PR TITLE
Add PodLocator API for topology-aware service routing

### PR DESCRIPTION
**Pull request by m1093782566**
**What type of PR is this?**

/kind api-change

/kind feature

**What this PR does / why we need it**:

Add PodLocator API for topology-aware service routing feature.

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

Design document of this feature is kubernetes/enhancements#640

**Does this PR introduce a user-facing change?**:

```release-note
Add PodLocator API
```

### CODE REVIEW

**Comment by liggitt** [comment]
linking PR's that had a test flake because of https://github.com/kubernetes/kubernetes/issues/71696

(the printing integration test failure is legitimate... the added type is missing a server-side printer. that said, I thought this was going to be done as a CRD at least for the initial experimentation)

File: pkg/apis/topology/types.go [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+type PodLocator struct {
+	metav1.TypeMeta
+	// +optional
+	metav1.ObjectMeta
+	// IP addresses of the running Pod.
+	// May not be loopback (127.0.0.0/8), link-local (169.254.0.0/16),
+	// or link-local multicast ((224.0.0.0/24).
+	// Both IPv4 and IPv6 are also accepted.
+	// +optional
+	Addresses []string
```
Comment (thockin) [inline pkg/apis/topology/types.go]: The dual-stack KEP proposes an IP struct here with metadata, but it's not clear what that metadata will be.  I'm inclined to request that we define struct here, even if it only holds one string field.  E.g.

```
    Addresses []IPInfo
```

and

```
type IPInfo struct {
  ip sting
}
```

**Review by thockin — COMMENTED** [review]
The biggest comment is that this should be done via CRD, unless we have a really really good reason why it can't.   @saad-ali has walked this path and has offered to help guide this one.

See #72046 - we should not merge unimplemented API changes. I would suggest leaving this open, and building other PRs on top of it, or even folding them into a single PR as well-commented commits in a single PR

**Comment by m1093782566** [comment]
Thanks for your review comments @thockin @liggitt 

I am open to implement this feature via CRD. I will ask @saad-ali for guidance if needed. 

THANKS!

File: pkg/apis/topology/types.go (hunk 1 above)
Comment (thockin) [reply pkg/apis/topology/types.go]: Ping.   If this is headed towards implementation, I'll need to get on reviews.

Comment (m1093782566) [reply pkg/apis/topology/types.go]: Sorry for delay @thockin 

I had few bandwidth in 1.14 release cycle and will have much more bandwidth in 1.15 bandwidth. Anyway, we will make it happen in 1.15. Thanks!

**Comment by m1093782566** [comment]
I am going to close this for the sake of focusing on #72046

File: pkg/apis/topology/types.go [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+	// Name of Node where the Pod is running on.
+	// +optional
+	NodeName string
+	// Labels of Node where the Pod is running on, being added for identifying Pod's topological information.
+	// +optional
+	NodeLabels map[string]string
```
Comment (thockin) [inline pkg/apis/topology/types.go]: I went back thru the notes from the big discussion.  Th eway I read it, we agreed NOT to have this field yet.

For alpha, we can make kube-proxy watch nodes.  In fact I have a partial PR for that already.  Then we have time to figure out if we can have proper metadata-only watches or if we need to add this or if we need to pre-cook to NodeMeta.

Comment (m1093782566) [reply pkg/apis/topology/types.go]: So let me just clarify:

We still need the `PodLocator` CRD and kube-dns/coredns will make use of it, right?

**Comment by johnbelamaric** [comment]
@thockin, if we watch nodes in kube-proxy, then we don't need pod locator
for kube-proxy because endpointaddress already has nodeName.

We do need it for DNS so we can do the source IP -> Pod -> nodeName. But if
we are watching nodes, we do not need the node labels on PodLocator.

The two paths are to get to *clusterIP* services support in alpha:
- kube-proxy watches Nodes, we do not need podlocator at all. Everything we
need is in Endpoints & Nodes.
- kube-proxy does not watch Nodes. we add podlocator, and it must include
node labels (otherwise we need to watch nodes!)

The second one sets us up for supporting headless services too - the first
does not, but it's a smaller change for now.

John

**Pull request closed without merging**